src/main/java/com/minecolonies/coremod/colony/buildings/modules/FieldsModule.java

```diff
@@ -0,0 +1,245 @@
+     *
+     * @return an integer stating the maximum field count.
+     */
+    protected abstract int getMaxFieldCount();
+
+    /**
+     * Get the class type which is expected for the fields to have.
+     *
+     * @return the class type.
+     */
+    public abstract Class<?> getExpectedFieldType();
+
+    /**
+     * Getter of the current field.
+     *
+     * @return a field object.
+     */
+    @Nullable
+    public IField getCurrentField()
+    {
+        return currentField;
+    }
+
+    /**
+     * Retrieves the field to work on for the citizen, as long as the current field has work, it will keep returning that field.
+     * Else it will retrieve a random field to work on for the citizen.
+     * This method will also automatically claim any fields that are not in use if the building is on automatic assignment mode.
+     *
+     * @return a field to work on.
+     */
+    @Nullable
+    public IField getFieldToWorkOn()
+    {
+        if (currentField != null && currentField.needsWork())
+        {
+            return currentField;
+        }
+
+        final List<IField> randomizedFields = getOwnedFields().stream()
+                                                .filter(field -> !field.equals(currentField))
+                                                .collect(CollectorUtils.toShuffledList());
+        for (final IField field : randomizedFields)
+        {
+            if (field.needsWork())
+            {
+                currentField = field;
+                return field;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Returns list of owned fields.
+     *
+     * @return a list of field objects.
+     */
+    @NotNull
+    public final List<IField> getOwnedFields()
+    {
+        return getFields().stream().filter(f -> building.getID().equals(f.getBuildingId())).toList();
+    }
+
+    /**
+     * Returns list of fields.
+     *
+     * @return a list of field objects.
+     */
+    @NotNull
+    public abstract List<IField> getFields();
+
+    /**
+     * Attempt to automatically claim free fields, if possible and if any fields are available.
+     */
+    public void claimFields()
+    {
+        if (!shouldAssignManually)
+        {
+            for (IField field : getFreeFields())
+            {
+                assignField(field);
+            }
+        }
+    }
+
+    /**
+     * Returns list of free fields.
+     *
+     * @return a list of field objects.
+     */
+    public final List<IField> getFreeFields()
+    {
+        return getFields().stream().filter(field -> !field.isTaken()).toList();
+    }
+
+    /**
+     * Method called to assign a field to the building.
+     *
+     * @param field the field to add.
+     */
+    public void assignField(final IField field)
+    {
+        if (canAssignField(field))
+        {
+            building.getColony().getBuildingManager().getField(otherField -> otherField.equals(field)).ifPresent(f -> {
```

Comment:
The reason it's done like this is because the field instance provided here usually comes from serialized sources.
Meaning that this input field value is not actually tied to a field as present in the building manager, if I were to update that instance directly, nothing would happen.